test/gurka/gurka.cc

```diff
@@ -117,7 +117,32 @@ std::string build_valhalla_request(const std::string& location_type,
 
   // we do this last so that options are additive/overwrite
   for (const auto& kv : options) {
-    rapidjson::Pointer(kv.first).Set(doc, kv.second);
+    // avoid polygons needs to be turned into an array
+    if (kv.second.find("[") != std::string::npos) {
+      rapidjson::Document d;
+      d.Parse(kv.second);
+
+      // Workaround start
+      rapidjson::Value a(rapidjson::kArrayType);
+      for (const auto& poly : d.GetArray()) {
+        rapidjson::Value a1(rapidjson::kArrayType);
+        for (const auto& coords : poly.GetArray()) {
+          rapidjson::Value a2(rapidjson::kArrayType);
+          for (const auto& v : coords.GetArray()) {
+            a2.PushBack(v.GetDouble(), allocator);
+          }
+          a1.PushBack(a2, allocator);
+        }
+        a.PushBack(a1, allocator);
+      }
+      // Workaround end
+      // setting d directly somehow messes up the very first coordinate's X... bug in rapidjson? can't
+      // reproduce in isolated test..
+      // rapidjson::SetValueByPointer(doc, "/avoid_polygons", d.GetArray());
+      rapidjson::Pointer(kv.first).Set(doc, a);
+    } else {
+      rapidjson::Pointer(kv.first).Set(doc, kv.second);
+    }
```

Comment:
disgusting I know.. I tried the conventional rapidjson way, what happened then was that it for some weird reason turned a coordinate string of `[[[13.51031, 52.234141], [13.53159, 52.24115]...]]` into `[[[-300e-52, 52.234141], [13.53159, 52.24115]...]]`, i.e. the very first X coord got completely messed up.. I tried to debug that for half a day and eventually did this abomination..

Either I was doing smth really stupid or it's a bug in rapidjson. Maybe someone more familiar with rapidjson can have a look? (@mandeepsandhu or @danpat ?)

Reply:
i think in the calling code you have to put something like 

```c++
{
  {"/avoid_polygons/0/0/0", lon0},
  {"/avoid_polygons/0/0/1", lat0},
  {"/avoid_polygons/0/1/0", lon1},
  {"/avoid_polygons/0/1/1", lat1},
  ...
}
```

Reply:
of course :facepalm: